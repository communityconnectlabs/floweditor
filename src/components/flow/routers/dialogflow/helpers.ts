import { NodeEditorSettings } from 'store/nodeEditor';

import { Types } from 'config/interfaces';
import { CallDialogflow } from 'flowTypes';
import { AssetType, RenderNode } from 'store/flowContext';
import { getActionUUID } from 'components/flow/actions/helpers';
import { createWebhookBasedNode } from '../helpers';
import { DialogflowRouterFormState } from './DialogflowRouterForm';

export const nodeToState = (settings: NodeEditorSettings): DialogflowRouterFormState => {
  const nodeFirstAction = settings.originalAction || settings.originalNode.node.actions[0];
  if (nodeFirstAction && nodeFirstAction.type === Types.call_dialogflow) {
    const action = nodeFirstAction as CallDialogflow;
    return {
      resultName: { value: action.result_name },
      dialogflowDB: {
        value: {
          id: action.dialogflow_db.id,
          name: action.dialogflow_db.text,
          type: AssetType.Dialogflow
        }
      },
      questionSrc: { value: action.question_src },
      valid: true
    };
  }
  return {
    resultName: { value: 'Result' },
    questionSrc: { value: '' },
    dialogflowDB: {
      value: { id: '', name: '', type: AssetType.Dialogflow }
    },
    valid: false
  };
};

export const stateToNode = (
  nodeSettings: NodeEditorSettings,
  formState: DialogflowRouterFormState
): RenderNode => {
  const action: CallDialogflow = {
    type: Types.call_dialogflow,
    result_name: formState.resultName.value,
    dialogflow_db: {
      id: formState.dialogflowDB.value.id,
      // @ts-ignore
      text: formState.dialogflowDB.value.name || formState.dialogflowDB.value.text
    },
    question_src: formState.questionSrc.value,
    uuid: getActionUUID(nodeSettings, Types.call_dialogflow)
  };

  return createWebhookBasedNode(action, nodeSettings.originalNode, false);
};
