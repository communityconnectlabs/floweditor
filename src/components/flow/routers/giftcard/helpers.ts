import { getActionUUID } from 'components/flow/actions/helpers';
import { CallGiftcard } from 'flowTypes';
import { Types } from 'config/interfaces';
import { NodeEditorSettings } from 'store/nodeEditor';
import { GiftCardRouterFormState } from './GiftCardRouterForm';
import { AssetType, RenderNode } from 'store/flowContext';
import { createWebhookBasedNode } from '../helpers';

export const nodeToState = (settings: NodeEditorSettings): GiftCardRouterFormState => {
  const nodeFirstAction = settings.originalAction || settings.originalNode.node.actions[0];
  if (nodeFirstAction && nodeFirstAction.type === Types.call_giftcard) {
    const action = nodeFirstAction as CallGiftcard;
    return {
      resultName: { value: action.result_name },
      giftcardDb: {
        value: {
          id: action.giftcard_db.id,
          name: action.giftcard_db.text,
          type: AssetType.GiftCard
        }
      },
      giftcardType: { value: action.giftcard_type },
      valid: true
    };
  }
  return {
    resultName: { value: 'Result' },
    giftcardDb: {
      value: { id: '', name: '', type: AssetType.GiftCard }
    },
    giftcardType: { value: 'GIFTCARD_ASSIGNING' },
    valid: false
  };
};

export const stateToNode = (
  nodeSettings: NodeEditorSettings,
  formState: GiftCardRouterFormState
): RenderNode => {
  const action: CallGiftcard = {
    type: Types.call_giftcard,
    result_name: formState.resultName.value,
    giftcard_db: {
      id: formState.giftcardDb.value.id,
      // @ts-ignore
      text: formState.giftcardDb.value.name || formState.giftcardDb.value.text
    },
    giftcard_type: formState.giftcardType.value,
    uuid: getActionUUID(nodeSettings, Types.call_giftcard)
  };

  return createWebhookBasedNode(action, nodeSettings.originalNode, false);
};
