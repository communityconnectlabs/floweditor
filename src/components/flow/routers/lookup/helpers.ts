import { createWebhookBasedNode } from 'components/flow/routers/helpers';
import { LookupRouterFormState } from 'components/flow/routers/lookup/LookupRouterForm';
import { Types } from 'config/interfaces';
import { CallLookup, LookupField } from 'flowTypes';
import { RenderNode, AssetType } from 'store/flowContext';
import { NodeEditorSettings, StringEntry, FormEntry } from 'store/nodeEditor';
import { createUUID } from 'utils';

export interface LookupQueryEntry extends FormEntry {
  field: StringEntry;
  rule: StringEntry;
  value: StringEntry;
}

export const getOriginalAction = (settings: NodeEditorSettings): CallLookup => {
  const action = settings.originalAction || settings.originalNode.node.actions[0];

  if (action && action.type === Types.call_lookup) {
    return action as CallLookup;
  }
};

export const nodeToState = (settings: NodeEditorSettings): LookupRouterFormState => {
  const resultName: StringEntry = { value: 'Result' };

  const state: LookupRouterFormState = {
    lookupDb: {
      value: { id: '', name: '', type: AssetType.Lookup }
    },
    lookupQueries: [],
    resultName,
    valid: false
  };

  if (settings.originalNode.ui.type === Types.split_by_lookup) {
    const action = getOriginalAction(settings) as CallLookup;

    state.resultName = { value: action.result_name };
    state.lookupDb = {
      value: {
        id: action.lookup_db.id,
        name: action.lookup_db.text,
        type: AssetType.Lookup
      }
    };
    state.lookupQueries = action.lookup_queries;
    state.valid = true;
  }

  return state;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: LookupRouterFormState
): RenderNode => {
  let uuid = createUUID();

  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const newAction: CallLookup = {
    uuid,
    lookup_queries: state.lookupQueries,
    type: Types.call_lookup,
    lookup_db: {
      id: state.lookupDb.value.id,
      text: state.lookupDb.value.name
    },
    result_name: state.resultName.value
  };
  console.log(newAction);
  return createWebhookBasedNode(newAction, settings.originalNode);
};
