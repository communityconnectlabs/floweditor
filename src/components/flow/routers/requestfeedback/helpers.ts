import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { RequestFeedback, Router, RouterTypes, SwitchRouter, Wait, WaitTypes } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';
import { RequestFeedbackFormState } from './RequestFeedbackForm';
import { RenderNode } from 'store/flowContext';
import { createRenderNode, resolveRoutes } from 'components/flow/routers/helpers';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { getType } from 'config/typeConfigs';

export const initializeForm = (settings: NodeEditorSettings): RequestFeedbackFormState => {
  if (settings.originalNode && getType(settings.originalNode) === Types.request_feedback) {
    const action = settings.originalNode.node.actions[0] as RequestFeedback;
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      let timeout = 0;
      if (router.wait && router.wait.timeout) {
        timeout = router.wait.timeout.seconds || 0;
      }
      return {
        starRatingQuestion: { value: action.star_rating_question },
        commentQuestion: { value: action.comment_question },
        resultName: { value: router.result_name || '' },
        timeout: timeout,
        valid: true
      };
    }
  }

  return {
    starRatingQuestion: { value: '' },
    commentQuestion: { value: '' },
    resultName: { value: 'Result' },
    timeout: 0,
    valid: false
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: RequestFeedbackFormState
): RenderNode => {
  const { cases, exits, defaultCategory, timeoutCategory, caseConfig, categories } = resolveRoutes(
    [],
    state.timeout > 0,
    settings.originalNode.node
  );

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const wait = { type: WaitTypes.msg } as Wait;
  if (state.timeout > 0) {
    wait.timeout = {
      seconds: state.timeout,
      category_uuid: timeoutCategory
    };
  }

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: defaultCategory,
    cases,
    categories,
    operand: DEFAULT_OPERAND,
    wait,
    ...optionalRouter
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.request_feedback,
    [stateToAction(settings, state)],
    { cases: caseConfig }
  );

  return newRenderNode;
};

export const stateToAction = (
  settings: NodeEditorSettings,
  state: RequestFeedbackFormState
): RequestFeedback => ({
  uuid: getActionUUID(settings, Types.request_feedback),
  type: Types.request_feedback,
  star_rating_question: state.starRatingQuestion.value,
  comment_question: state.commentQuestion.value
});
