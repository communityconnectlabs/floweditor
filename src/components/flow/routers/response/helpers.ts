/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import {
  createCaseProps,
  createRenderNode,
  hasCases,
  resolveRoutes
} from 'components/flow/routers/helpers';
import { ResponseRouterFormState } from 'components/flow/routers/response/ResponseRouterForm';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Types, Operators } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { Router, RouterTypes, SwitchRouter, Wait, WaitTypes } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { SelectOption } from 'components/form/select/SelectElement';
import { getLocalizations } from 'store/helpers';
import { RouterFormProps } from 'components/flow/props';

export enum AutomatedTestCaseType {
  AUTO_GENERATED,
  USER_GENERATED
}

export const ALLOWED_USER_TESTS = [
  Operators.has_email,
  Operators.has_phone,
  Operators.has_pattern,
  Operators.has_date,
  Operators.has_date_lt,
  Operators.has_date_eq,
  Operators.has_date_gt,
  Operators.has_time,
  Operators.has_text,
  Operators.has_number,
  Operators.has_number_lt,
  Operators.has_number_lte,
  Operators.has_number_eq,
  Operators.has_number_gte,
  Operators.has_number_gt,
  Operators.has_number_between
];

export const ALLOWED_AUTO_TESTS = [
  Operators.has_any_word,
  Operators.has_all_words,
  Operators.has_phrase,
  Operators.has_only_phrase,
  Operators.has_beginning
];

export const ALLOWED_TESTS = [...ALLOWED_USER_TESTS, ...ALLOWED_AUTO_TESTS];

export interface AutomatedTestCase {
  type: AutomatedTestCaseType;
  testText: string;
  actualCategory: string;
  confirmedCategory: string;
  categoriesMatch: boolean;
  confirmed: boolean;
  deleted: boolean;
}

interface ConfigRouter {
  spell_checker?: boolean;
  spelling_correction_sensitivity?: string;
  test_cases?: { [lang: string]: AutomatedTestCase[] };
}

export const matchResponseTextWithCategory = (text: string, cases: CaseProps[]): string[] => {
  let matches: string[] = [];
  let args: string[] = [];
  let emailRegExp = /.*\b(?<email>\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+)\b.*/;
  let phoneRegExp = /.*\b(?<phone>\+?(?:[0-9] ?){6,14}[0-9])\b.*/;
  let dateRegExp = /.*\b(?<date>([0-9]{1,2})[-.\\/_]([0-9]{1,2})[-.\\/_]([0-9]{4}|[0-9]{2})|([0-9]{4})[-.\\/_]([0-9]{1,2})[-.\\/_]([0-9]{1,2})|\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(\.(\d{0,9}))?([+-]\d{2}:\d{2}|Z))\b.*/;
  let timeRegExp = /.*\b(?<time>([0-9]{1,2}):([0-9]{2})(:([0-9]{2})(\.(\d+))?)?\W*([aApP][mM])?)\b.*/;
  let numberRegExp = /.*\b(?<number>[$£€]?([\d,][\d,.]*([.,]\d+)?)\D*$)\b.*/;
  let originalText = text;
  text = text.toLowerCase();
  cases.some(item => {
    let match = false;
    let type = item.kase.type;
    switch (type) {
      case 'has_any_word':
        args = item.kase.arguments[0].toLowerCase().split(/\s+/);
        args = args.filter(arg => arg !== '');
        match = args.some(element => new RegExp('(' + element + ')').test(text));
        break;
      case 'has_all_words':
        args = item.kase.arguments[0].toLowerCase().split(/\s+/);
        args = args.filter(arg => arg !== '');
        match = args.every(element => new RegExp('(' + element + ')').test(text));
        break;
      case 'has_phrase':
        match = new RegExp('(^|.*\\s)(' + item.kase.arguments[0].toLowerCase() + ')(\\s.*|$)').test(
          text
        );
        break;
      case 'has_only_phrase':
        match = new RegExp('^(' + item.kase.arguments[0].toLowerCase() + ')$').test(text);
        break;
      case 'has_beginning':
        match = new RegExp('^(' + item.kase.arguments[0].toLowerCase() + ').*').test(text);
        break;
      case 'has_email':
        args = text.split(/\s+/);
        args = args.filter(arg => arg !== '');
        match = args.some(word => emailRegExp.test(word));
        break;
      case 'has_phone':
        match = phoneRegExp.test(text);
        break;
      case 'has_pattern':
        try {
          match = new RegExp(item.kase.arguments[0]).test(originalText);
          // eslint-disable-next-line no-empty
        } catch (error) {}
        break;
      case 'has_date':
        var msgDate = dateRegExp.exec(originalText);
        if (msgDate) {
          match = !isNaN(Date.parse(msgDate.groups.date));
        }
        break;
      case 'has_date_eq':
        var kwargs = /(?<days>-?\d+)/.exec(item.kase.arguments[0]);
        var days = Number.parseInt((kwargs ? kwargs : { groups: { days: '' } }).groups.days);
        // eslint-disable-next-line no-redeclare
        var msgDate = dateRegExp.exec(originalText);
        if (!isNaN(days) && msgDate && Date.parse(msgDate.groups.date)) {
          var exactDay = new Date();
          var testedDate = new Date(Date.parse(msgDate.groups.date));
          testedDate.setHours(0, 0, 0, 0);
          exactDay.setDate(exactDay.getDate() + days);
          exactDay.setHours(0, 0, 0, 0);
          match = exactDay.valueOf() === testedDate.valueOf();
        }
        break;
      case 'has_date_lt':
        // eslint-disable-next-line no-redeclare
        var kwargs = /(?<days>-?\d+)/.exec(item.kase.arguments[0]);
        // eslint-disable-next-line no-redeclare
        var days = Number.parseInt((kwargs ? kwargs : { groups: { days: '' } }).groups.days);
        // eslint-disable-next-line no-redeclare
        var msgDate = dateRegExp.exec(originalText);
        if (!isNaN(days) && msgDate && !isNaN(Date.parse(msgDate.groups.date))) {
          // eslint-disable-next-line no-redeclare
          var exactDay = new Date();
          // eslint-disable-next-line no-redeclare
          var testedDate = new Date(Date.parse(msgDate.groups.date));
          testedDate.setHours(0, 0, 0, 0);
          exactDay.setDate(exactDay.getDate() + days);
          exactDay.setHours(0, 0, 0, 0);
          match = exactDay.valueOf() > testedDate.valueOf();
        }
        break;
      case 'has_date_gt':
        // eslint-disable-next-line no-redeclare
        var kwargs = /(?<days>-?\d+)/.exec(item.kase.arguments[0]);
        // eslint-disable-next-line no-redeclare
        var days = Number.parseInt((kwargs ? kwargs : { groups: { days: '' } }).groups.days);
        // eslint-disable-next-line no-redeclare
        var msgDate = dateRegExp.exec(originalText);
        if (!isNaN(days) && msgDate && !isNaN(Date.parse(msgDate.groups.date))) {
          // eslint-disable-next-line no-redeclare
          var exactDay = new Date();
          // eslint-disable-next-line no-redeclare
          var testedDate = new Date(Date.parse(msgDate.groups.date));
          testedDate.setHours(0, 0, 0, 0);
          exactDay.setDate(exactDay.getDate() + days);
          exactDay.setHours(0, 0, 0, 0);
          match = exactDay.valueOf() < testedDate.valueOf();
        }
        break;
      case 'has_time':
        match = timeRegExp.test(text);
        break;
      case 'has_text':
        match = Boolean(originalText.trim());
        break;
      case 'has_number':
        match = numberRegExp.test(text);
        break;
      case 'has_number_lt':
        var testingNum = numberRegExp.exec(text);
        if (testingNum) {
          match =
            Number.parseFloat(item.kase.arguments[0]) > Number.parseFloat(testingNum.groups.number);
        }
        break;
      case 'has_number_lte':
        // eslint-disable-next-line no-redeclare
        var testingNum = numberRegExp.exec(text);
        if (testingNum) {
          match =
            Number.parseFloat(item.kase.arguments[0]) >=
            Number.parseFloat(testingNum.groups.number);
        }
        break;
      case 'has_number_eq':
        // eslint-disable-next-line no-redeclare
        var testingNum = numberRegExp.exec(text);
        if (testingNum) {
          match =
            Number.parseFloat(item.kase.arguments[0]) ===
            Number.parseFloat(testingNum.groups.number);
        }
        break;
      case 'has_number_gte':
        // eslint-disable-next-line no-redeclare
        var testingNum = numberRegExp.exec(text);
        if (testingNum) {
          match =
            Number.parseFloat(item.kase.arguments[0]) <=
            Number.parseFloat(testingNum.groups.number);
        }
        break;
      case 'has_number_gt':
        // eslint-disable-next-line no-redeclare
        var testingNum = numberRegExp.exec(text);
        if (testingNum) {
          match =
            Number.parseFloat(item.kase.arguments[0]) < Number.parseFloat(testingNum.groups.number);
        }
        break;
      case 'has_number_between':
        // eslint-disable-next-line no-redeclare
        var testingNum = numberRegExp.exec(text);
        if (testingNum) {
          match =
            Number.parseFloat(item.kase.arguments[0]) <
              Number.parseFloat(testingNum.groups.number) &&
            Number.parseFloat(item.kase.arguments[1]) > Number.parseFloat(testingNum.groups.number);
        }
        break;
      default:
        break;
    }
    if (match && item.categoryName !== '') {
      matches.push(item.categoryName);
      // stop iterations to return first matched category
      return true;
    }
    return false;
  });
  return matches;
};

export const generateAutomatedTest = (
  caseItem: CaseProps,
  cases: CaseProps[]
): AutomatedTestCase => {
  let testCase = {
    type: AutomatedTestCaseType.AUTO_GENERATED,
    testText: caseItem.kase.arguments[0],
    actualCategory: matchResponseTextWithCategory(caseItem.kase.arguments[0], cases).join(', '),
    confirmedCategory: caseItem.categoryName
  };
  return testCase as AutomatedTestCase;
};

export const generateAutomatedTests = (cases: CaseProps[]): AutomatedTestCase[] => {
  let testCases: AutomatedTestCase[] = [];
  cases = cases.filter(case_ => ALLOWED_AUTO_TESTS.includes(case_.kase.type));
  cases.forEach(item => {
    let testCase = generateAutomatedTest(item, cases);
    if (testCase.confirmedCategory) {
      testCases.push(testCase as AutomatedTestCase);
    }
  });
  return testCases;
};

export const getLocalizedCases = (cases: CaseProps[], props: RouterFormProps, isoLang: string) => {
  if (isoLang === props.nodeSettings.defaultLanguage) {
    return cases;
  }
  let localizations = getLocalizations(
    props.nodeSettings.originalNode.node,
    null,
    props.assetStore.languages.items[isoLang],
    props.nodeSettings.localization[isoLang]
  );
  // eslint-disable-next-line no-undef
  let localizedCases = new Map();
  localizations.forEach(localization => {
    let localizedObj = localization.getObject();
    if (
      localizedObj.hasOwnProperty('uuid') &&
      localizedObj.hasOwnProperty('type') &&
      localizedObj.hasOwnProperty('category_uuid') &&
      localizedObj.hasOwnProperty('arguments')
    ) {
      localizedCases.set(localizedObj.uuid, localizedObj);
    }
  });
  /* copying of cases to prevent updating of original cases */
  cases = JSON.parse(JSON.stringify(cases));
  cases.forEach(case_ => {
    let translatedKase = localizedCases.get(case_.kase.uuid);
    if (translatedKase) {
      case_.kase = translatedKase;
    }
  });
  return cases;
};

export const nodeToState = (settings: NodeEditorSettings, props?: any): ResponseRouterFormState => {
  let initialCases: CaseProps[] = [];

  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: 'Result' };
  let timeout = 0;
  let enabledSpell = false;
  let spellSensitivity = '70';
  let languages: SelectOption[] = [];
  let localizedCases: CaseProps[] = [];
  let currentLanguage: any = null;
  let testCases: any = {};

  if (settings.originalNode && getType(settings.originalNode) === Types.wait_for_response) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      if (hasCases(settings.originalNode.node)) {
        initialCases = createCaseProps(router.cases, settings.originalNode);
      }

      resultName = { value: router.result_name || '' };
    }

    if (router.config && router.config.spell_checker) {
      enabledSpell = router.config.spell_checker;
      spellSensitivity = router.config.spelling_correction_sensitivity;
    }

    if (settings.originalNode.node.router.wait && settings.originalNode.node.router.wait.timeout) {
      timeout = settings.originalNode.node.router.wait.timeout.seconds || 0;
    }

    Object.entries(props.assetStore.languages.items).forEach(([_, item]) => {
      // @ts-ignore
      languages.push({ label: item.name, value: item.id });
      // @ts-ignore
      if (item.id === (settings.defaultLanguage || 'eng')) {
        // @ts-ignore
        currentLanguage = { label: item.name, value: item.id };
      }
    });

    localizedCases = getLocalizedCases(initialCases, props, currentLanguage.value);

    if (router.config && router.config.test_cases) {
      testCases = router.config.test_cases;
      languages.forEach(lang => {
        if (!testCases[lang.value]) {
          let cases = getLocalizedCases(initialCases, props, lang.value);
          testCases[lang.value] = generateAutomatedTests(cases);
        }
      });
    } else {
      testCases[currentLanguage.value] = generateAutomatedTests(localizedCases);
      languages.forEach(lang => {
        let cases = getLocalizedCases(initialCases, props, lang.value);
        testCases[lang.value] = generateAutomatedTests(cases);
      });
    }
  }

  if (!currentLanguage) {
    currentLanguage = { label: 'English', value: settings.defaultLanguage || 'eng' };
    testCases[currentLanguage.value] = [];
    languages.push(currentLanguage);
  }

  return {
    cases: initialCases,
    resultName,
    timeout,
    enabledSpell,
    spellSensitivity,
    valid: true,
    testingLangs: languages,
    testingLang: currentLanguage,
    liveTestText: { value: '' },
    automatedTestCases: testCases,
    localizedCases
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ResponseRouterFormState
): RenderNode => {
  const { cases, exits, defaultCategory, timeoutCategory, caseConfig, categories } = resolveRoutes(
    state.cases,
    state.timeout > 0,
    settings.originalNode.node
  );

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const wait = { type: WaitTypes.msg } as Wait;
  if (state.timeout > 0) {
    wait.timeout = {
      seconds: state.timeout,
      category_uuid: timeoutCategory
    };
  }

  const config: ConfigRouter = {};
  if (state.enabledSpell) {
    config.spell_checker = state.enabledSpell;
    config.spelling_correction_sensitivity = state.spellSensitivity;
  }

  if (state.automatedTestCases) {
    config.test_cases = {};
    for (const [lang, testCases] of Object.entries(state.automatedTestCases)) {
      config.test_cases[lang] = testCases.filter(test => !test.deleted);
    }
  }

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: defaultCategory,
    cases,
    categories,
    operand: DEFAULT_OPERAND,
    wait,
    config,
    ...optionalRouter
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.wait_for_response,
    [],
    { cases: caseConfig }
  );

  return newRenderNode;
};
