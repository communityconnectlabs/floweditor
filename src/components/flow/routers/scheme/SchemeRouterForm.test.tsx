import { render, mock, fireEvent, getUpdatedNode, fireTembaSelect } from 'test/utils';
import * as React from 'react';
import * as utils from 'utils';
import SchemeRouterForm from './SchemeRouterForm';
import { createSchemeRouter, getRouterFormProps } from 'testUtils/assetCreators';
import { SCHEMES } from 'config/typeConfigs';
import { getSwitchRouter } from '../helpers';
import { Operators } from 'config/interfaces';

mock(utils, 'createUUID', utils.seededUUIDs());

const routerProps = getRouterFormProps(createSchemeRouter(SCHEMES.slice(0, 3)));

describe(SchemeRouterForm.name, () => {
  it('should render', () => {
    const { baseElement } = render(<SchemeRouterForm {...routerProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should reuse ids on updates', () => {
    const { getByText } = render(<SchemeRouterForm {...routerProps} />);
    fireEvent.click(getByText('Ok'));

    expect(getUpdatedNode(routerProps).node).toEqual(routerProps.nodeSettings.originalNode.node);
  });

  it('should select schemes', () => {
    const { getByTestId, getByText } = render(<SchemeRouterForm {...routerProps} />);

    fireTembaSelect(getByTestId('temba_select_channel_type'), 'whatsapp');
    fireEvent.click(getByText('Ok'));

    const router = getSwitchRouter(getUpdatedNode(routerProps).node);
    expect(router.cases.length).toBe(1);
    expect(router.cases[0].arguments[0]).toBe('whatsapp');
    expect(router.cases[0].type).toBe(Operators.has_only_phrase);
    expect(routerProps.updateRouter).toMatchCallSnapshot();
  });
});
