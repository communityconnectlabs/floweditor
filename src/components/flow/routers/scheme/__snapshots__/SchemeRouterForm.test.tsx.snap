// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemeRouterForm should render 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header split_by_scheme"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split by URN Type
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                searchable="true"
              >
                <temba-option
                  name="Send the contact a message"
                  value="send_msg"
                />
                <temba-option
                  name="Wait for the contact to respond"
                  value="wait_for_response"
                />
                <temba-option
                  name="Send somebody else a message"
                  value="send_broadcast"
                />
                <temba-option
                  name="Label the incoming message"
                  value="add_input_labels"
                />
                <temba-option
                  name="Add a URN for the contact"
                  value="add_contact_urn"
                />
                <temba-option
                  name="Add the contact to a group"
                  value="add_contact_groups"
                />
                <temba-option
                  name="Remove the contact from a group"
                  value="remove_contact_groups"
                />
                <temba-option
                  name="Update the contact"
                  value="set_contact_field"
                />
                <temba-option
                  name="Send an email"
                  value="send_email"
                />
                <temba-option
                  name="Save a result for this flow"
                  value="set_run_result"
                />
                <temba-option
                  name="Call a webhook"
                  value="call_webhook"
                />
                <temba-option
                  name="Call Zapier"
                  value="call_resthook"
                />
                <temba-option
                  name="Enter another flow"
                  value="enter_flow"
                />
                <temba-option
                  name="Start somebody else in a flow"
                  value="start_session"
                />
                <temba-option
                  name="Open a ticket with a human agent"
                  value="open_ticket"
                />
                <temba-option
                  name="Send the contact airtime"
                  value="transfer_airtime"
                />
                <temba-option
                  name="Split by intent"
                  value="split_by_intent"
                />
                <temba-option
                  name="Split by a custom expression"
                  value="split_by_expression"
                />
                <temba-option
                  name="Split by a contact field"
                  value="split_by_contact_field"
                />
                <temba-option
                  name="Split by a result in the flow"
                  value="split_by_run_result"
                />
                <temba-option
                  name="Split by random chance"
                  value="split_by_random"
                />
                <temba-option
                  name="Split by group membership"
                  value="split_by_groups"
                />
                <temba-option
                  name="Split by URN type"
                  selected="true"
                  value="split_by_scheme"
                />
              </temba-select>
            </div>
          </div>
        </div>
        <p>
          The contact's URN is the address they used to reach you such as their phone number or a Facebook ID. Select which URN types to split by below.
        </p>
        <div
          class="ele"
        >
          <div
            class="normal "
          >
            <temba-select
              data-testid="temba_select_channel_type"
              multi="true"
              placeholder="Select the channels to split by"
            >
              <temba-option
                name="SMS"
                selected="true"
                value="tel"
              />
              <temba-option
                name="WhatsApp"
                selected="true"
                value="whatsapp"
              />
              <temba-option
                name="Facebook"
                selected="true"
                value="facebook"
              />
              <temba-option
                name="Twitter"
                value="twitterid"
              />
              <temba-option
                name="Telegram"
                value="telegram"
              />
              <temba-option
                name="Viber"
                value="viber"
              />
              <temba-option
                name="Line"
                value="line"
              />
              <temba-option
                name="Wechat"
                value="wechat"
              />
              <temba-option
                name="Firebase"
                value="fcm"
              />
              <temba-option
                name="Jiochat"
                value="jiochat"
              />
              <temba-option
                name="Freshchat"
                value="freshchat"
              />
              <temba-option
                name="External"
                value="ext"
              />
            </temba-select>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <span
            class="toggle_link"
            data-spec="toggle-link"
          >
            Save as..
          </span>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`SchemeRouterForm should select schemes 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
        },
        Object {
          "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
        },
        Object {
          "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        },
        Object {
          "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "tel",
            ],
            "category_uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
            "type": "has_only_phrase",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          Object {
            "arguments": Array [
              "whatsapp",
            ],
            "category_uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
            "type": "has_only_phrase",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "arguments": Array [
              "facebook",
            ],
            "category_uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
            "type": "has_only_phrase",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
            "name": "SMS",
            "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          },
          Object {
            "exit_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
            "name": "WhatsApp",
            "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
          },
          Object {
            "exit_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
            "name": "Facebook",
            "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
          },
          Object {
            "exit_uuid": "6c22884b-0e35-4206-982e-18320691eda9",
            "name": "Other",
            "uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
          },
        ],
        "default_category_uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
        "operand": "@(urn_parts(contact.urn).scheme)",
        "result_name": "",
        "type": "switch",
      },
      "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
    },
    "ui": Object {
      "config": Object {
        "cases": Object {},
      },
      "position": null,
      "type": "split_by_scheme",
    },
  },
]
`;
